### PR TITLE
Add softfailure for bug: 1221812

### PR DESCRIPTION
**Pull request by RoyCai7**
Progress ticket: https://progress.opensuse.org/issues/163415
Bug: https://bugzilla.suse.com/show_bug.cgi?id=1221812

- Related ticket: https://progress.opensuse.org/issues/163415
- Needles: n/a
- Verification run: 
sles15sp6
http://openqa.qam.suse.cz/tests/77864#step/parallel_guest_migration_source/238
other version:
http://openqa.qam.suse.cz/tests/77870

### CODE REVIEW

File: lib/parallel_guest_migration_base.pm [hunk 1]
```diff
@@ -1750,6 +1750,26 @@ sub check_peer_test_run {
...
 
+=head2 filter_migration_tests
+
+Filters migration tests based on specified criteria.
+
+=cut
+
+sub filter_migration_tests {
+    my ($self, @guest_migration_test) = @_;
+    my @filtered_tests;
```
Comment (waynechen55) [inline lib/parallel_guest_migration_base.pm:1761]: Could you help switch to named argument ?

File: tests/virt_autotest/parallel_guest_migration_source.pm [hunk 2]
```diff
@@ -163,6 +163,10 @@ sub guest_migration_test {
 
     $self->set_test_run_progress;
     my @guest_migration_test = split(/,/, get_var('GUEST_MIGRATION_TEST', ''));
+
+    # Remove 'virsh_live_native_p2p_manual_postcopy' if present and record a soft failure
+    @guest_migration_test = $self->filter_migration_tests(@guest_migration_test);
```
Comment (waynechen55) [inline tests/virt_autotest/parallel_guest_migration_source.pm:168]: You skip this test on all versions of host system ?

Comment (RoyCai7) [reply tests/virt_autotest/parallel_guest_migration_source.pm:168]: yes, that's the issue, I forgot it. thanks for your reminder. will improve it.

File: lib/parallel_guest_migration_base.pm [hunk 3]
```diff
@@ -1750,6 +1750,26 @@ sub check_peer_test_run {
...
+
+    foreach my $migration_criteria (@guest_migration_test) {
+        if ($migration_criteria eq 'virsh_live_native_p2p_manual_postcopy') {
+            record_soft_failure('bsc#1221812 - qemu coredump when virsh migrate postcopy + virsh migrate-postcopy on sle15sp6 kvm host.');
+        } else {
+            push @filtered_tests, $migration_criteria;
```
Comment (alice-suse) [inline lib/parallel_guest_migration_base.pm:1767]: Just curious -- how would you update code when bsc#1221812 is fixed?

Comment (RoyCai7) [reply lib/parallel_guest_migration_base.pm:1767]: Delete code! I think.

**Comment by RoyCai7** [comment]
Still debugging, checking.

**Comment by RoyCai7** [comment]
Complete development, and pending final VR test.

**Comment by guoxuguang** [comment]
LGTM

File: lib/parallel_guest_migration_base.pm [hunk 4]
```diff
@@ -697,6 +697,10 @@ sub initialize_test_result {
 
     $self->set_test_run_progress;
     my @_guest_migration_test = split(/,/, get_var('GUEST_MIGRATION_TEST', ''));
+    # Remove 'virsh_live_native_p2p_manual_postcopy' if present and record a soft failure
+    if (is_sle('=15-sp6')) {
+        @_guest_migration_test = $self->filter_migration_tests(@_guest_migration_test);
+    }
```
Comment (waynechen55) [inline lib/parallel_guest_migration_base.pm:703]: Why do you need this block in base module ?

Comment (RoyCai7) [reply lib/parallel_guest_migration_base.pm:703]: pass test result, or job fails

File: lib/parallel_guest_migration_base.pm [hunk 5]
```diff
@@ -1750,6 +1754,26 @@ sub check_peer_test_run {
...
 
+=head2 filter_migration_tests
+
+Filters migration tests based on specified criteria.
+
+=cut
+
+sub filter_migration_tests {
+    my ($self, @migration_tests) = @_;
+    my @guest_migration_test;
```
Comment (waynechen55) [inline lib/parallel_guest_migration_base.pm:1765]: I think you'd better use named argument.
```
my ($self, %args) = @_;
```

All subroutines use named arguments in this module.

File: tests/virt_autotest/parallel_guest_migration_source.pm [hunk 6]
```diff
@@ -163,6 +164,10 @@ sub guest_migration_test {
 
     $self->set_test_run_progress;
     my @guest_migration_test = split(/,/, get_var('GUEST_MIGRATION_TEST', ''));
+    # Remove 'virsh_live_native_p2p_manual_postcopy' if present and record a soft failure
+    if (is_sle('=15-sp6')) {
+        @guest_migration_test = $self->filter_migration_tests(@guest_migration_test);
+    }
```
Comment (waynechen55) [inline tests/virt_autotest/parallel_guest_migration_source.pm:170]: I think one line is enough:
```
@guest_migration_test = $self->filter_migration_tests(@guest_migration_test)  if (is_sle('=15-sp6'));
```

File: tests/virt_autotest/parallel_guest_migration_destination.pm [hunk 7]
```diff
@@ -133,6 +134,10 @@ sub guest_migration_test {
 
     $self->set_test_run_progress;
     my @guest_migration_test = split(/,/, get_var('GUEST_MIGRATION_TEST'));
+    # Remove 'virsh_live_native_p2p_manual_postcopy' if present and record a soft failure
+    if (is_sle('=15-sp6')) {
+        @guest_migration_test = $self->filter_migration_tests(@guest_migration_test);
+    }
```
Comment (waynechen55) [inline tests/virt_autotest/parallel_guest_migration_destination.pm:140]: I think one line is enough:
```
@guest_migration_test = $self->filter_migration_tests(@guest_migration_test)  if (is_sle('=15-sp6'));
```

File: lib/parallel_guest_migration_base.pm [hunk 8]
```diff
@@ -1750,6 +1754,26 @@ sub check_peer_test_run {
...
+
+    foreach my $migration_criteria (@migration_tests) {
+        if ($migration_criteria eq 'virsh_live_native_p2p_manual_postcopy') {
+            record_soft_failure('bsc#1221812 - qemu coredump when virsh migrate postcopy + virsh migrate-postcopy on sle15sp6 kvm host.');
+        } else {
+            push @guest_migration_test, $migration_criteria;
+        }
+    }
```
Comment (waynechen55) [inline lib/parallel_guest_migration_base.pm:1773]: I think you can save ```$migration_criteria ```. Just use ```$_``` inside the loop.

**Comment by alice-suse** [comment]
I guess it is ready for merge? @RoyCai7 please reply if there is other ongoing code rework or verification.

**Comment by RoyCai7** [comment]
> I guess it is ready for merge? @RoyCai7 please reply if there is other ongoing code rework or verification.

Completed development, and completed VR tests.
Verification run:
sles15sp6
http://openqa.qam.suse.cz/tests/77864#step/parallel_guest_migration_source/238
other version:
http://openqa.qam.suse.cz/tests/77870

**Comment by alice-suse** [comment]
Thank you, then let's merge it.

**Pull request merged**